crates/iceberg/src/writer/base_writer/equality_delete_writer.rs

```diff
@@ -0,0 +1,502 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module provide `EqualityDeleteWriter`.
+
+use std::sync::Arc;
+
+use arrow_array::RecordBatch;
+use arrow_schema::{DataType, Field, SchemaRef as ArrowSchemaRef};
+use itertools::Itertools;
+use parquet::arrow::PARQUET_FIELD_ID_META_KEY;
+
+use crate::arrow::record_batch_projector::RecordBatchProjector;
+use crate::arrow::schema_to_arrow_schema;
+use crate::spec::{DataFile, SchemaRef, Struct};
+use crate::writer::file_writer::{FileWriter, FileWriterBuilder};
+use crate::writer::{IcebergWriter, IcebergWriterBuilder};
+use crate::{Error, ErrorKind, Result};
+
+/// Builder for `EqualityDeleteWriter`.
+#[derive(Clone)]
+pub struct EqualityDeleteFileWriterBuilder<B: FileWriterBuilder> {
+    inner: B,
+}
+
+impl<B: FileWriterBuilder> EqualityDeleteFileWriterBuilder<B> {
+    /// Create a new `EqualityDeleteFileWriterBuilder` using a `FileWriterBuilder`.
+    pub fn new(inner: B) -> Self {
+        Self { inner }
+    }
+}
+
+/// Config for `EqualityDeleteWriter`.
+pub struct EqualityDeleteWriterConfig {
+    // Field ids used to determine row equality in equality delete files.
+    equality_ids: Vec<i32>,
+    // Projector used to project the data chunk into specific fields.
+    projector: RecordBatchProjector,
+    partition_value: Struct,
+}
+
+impl EqualityDeleteWriterConfig {
+    /// Create a new `DataFileWriterConfig` with equality ids.
+    pub fn new(
+        equality_ids: Vec<i32>,
+        original_schema: SchemaRef,
+        partition_value: Option<Struct>,
+    ) -> Result<Self> {
+        let original_arrow_schema = Arc::new(schema_to_arrow_schema(&original_schema)?);
+        let projector = RecordBatchProjector::new(
+            original_arrow_schema,
+            &equality_ids,
+            // The following rule comes from https://iceberg.apache.org/spec/#identifier-field-ids
+            // - The identifier field ids must be used for primitive types.
+            // - The identifier field ids must not be used for floating point types or nullable fields.
+            // - The identifier field ids can be nested field of struct but not nested field of nullable struct.
+            |field| {
+                // Only primitive type is allowed to be used for identifier field ids
+                if field.is_nullable()
+                    || !field.data_type().is_primitive()
+                    || matches!(
+                        field.data_type(),
+                        DataType::Float16 | DataType::Float32 | DataType::Float64
+                    )
```

Comment:
[This part decides whether this field is possible to be the identifier field](https://iceberg.apache.org/spec/#identifier-field-ids:~:text=Float%2C%20double%2C%20and%20optional%20fields%20cannot%20be%20used%20as%20identifier%20fields). 
searchable_field_func is used to decide whether we should search the struct. [We need this because a nested field cannot be used as an identifier field if it is nested in an optional struct](https://iceberg.apache.org/spec/#identifier-field-ids:~:text=a%20nested%20field%20cannot%20be%20used%20as%20an%20identifier%20field%20if%20it%20is%20nested%20in%20an%20optional%20struct).